### ISSUE TITLE
Unable to Install luisgen 

### ISSUE THREAD

**Issue by ponarunkumar** [Microsoft/botbuilder-tools#275]
**ERR! Failed at PostInstall Script**
Unable to install Luisgen in ubuntu (bionic beaver) 18.04
**To Reproduce**
1. sudo npm install luisgen -g

`vagrant@ubuntu-bionic:~$ sudo npm install luisgen -g
/usr/local/bin/luisgen -> /usr/local/lib/node_modules/luisgen/bin/luisgen.js

> dotnet-2.0.0@1.4.4 postinstall /usr/local/lib/node_modules/luisgen/node_modules/dotnet-2.0.0
> node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"

events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: spawn /usr/bin/node ENOENT
    at _errnoException (util.js:1022:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:190:19)
    at onErrorNT (internal/child_process.js:372:16)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! Linux 4.15.0-23-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "luisgen" "-g"
npm ERR! node v8.10.0
npm ERR! npm  v3.5.2
npm ERR! code ELIFECYCLE

npm ERR! dotnet-2.0.0@1.4.4 postinstall: `node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the dotnet-2.0.0@1.4.4 postinstall script 'node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the dotnet-2.0.0 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs dotnet-2.0.0
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls dotnet-2.0.0
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/vagrant/npm-debug.log
`

**Expected behavior**
successful installation

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04 LTS

**Comment by vishwacsena** [comment]
@tsuwandy - have we seen anything similar on dispatch? 
@chrimc62 FYI

**Comment by ponarunkumar** [comment]
@tsuwandy @chrimc62 Yes. similar issue with dispatch.
`> botdispatch@1.0.4 postinstall /usr/local/lib/node_modules/botdispatch
> npm install -g dotnet-2.0.0

npm ERR! Linux 4.15.0-23-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "-g" "dotnet-2.0.0"
npm ERR! node v8.10.0
npm ERR! npm  v3.5.2
npm ERR! path /home/vagrant/.npm/dotnet-2.0.0/1.4.4/package/package.json.1625320364
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall open

npm ERR! Error: EACCES: permission denied, open '/home/vagrant/.npm/dotnet-2.0.0/1.4.4/package/package.json.1625320364'
npm ERR!  { Error: EACCES: permission denied, open '/home/vagrant/.npm/dotnet-2.0.0/1.4.4/package/package.json.1625320364'
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: '/home/vagrant/.npm/dotnet-2.0.0/1.4.4/package/package.json.1625320364' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! Please include the following file with any support request:
npm ERR!     /usr/local/lib/node_modules/botdispatch/npm-debug.log
npm ERR! Linux 4.15.0-23-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "-g" "botdispatch"
npm ERR! node v8.10.0
npm ERR! npm  v3.5.2
npm ERR! code ELIFECYCLE

npm ERR! botdispatch@1.0.4 postinstall: `npm install -g dotnet-2.0.0`
npm ERR! Exit status 243
npm ERR! 
npm ERR! Failed at the botdispatch@1.0.4 postinstall script 'npm install -g dotnet-2.0.0'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the botdispatch package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm install -g dotnet-2.0.0
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs botdispatch
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls botdispatch
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/vagrant/npm-debug.log

**Comment by jwiley84** [comment]
I was able to 'sudo npm install luisgen -g' on 16.04 (xenial xerus). I'm upgrading to 18.04 to see if I can repro.

**Comment by ponarunkumar** [comment]
Thanks @jwiley84 , I will temporarily switch to 16.04 (xenial xerus) , However it would be great to have it in the current LTS 18.04 bionic.

**Comment by ponarunkumar** [comment]
@jwiley84 the issue persists in xenial xerus as well,. any alternative steps that you could suggest?
`vagrant@ubuntu-xenial:/$ sudo npm install -g luisgen
/usr/local/bin/luisgen -> /usr/local/lib/node_modules/luisgen/bin/luisgen.js

> dotnet-2.0.0@1.4.4 postinstall /usr/local/lib/node_modules/luisgen/node_modules/dotnet-2.0.0
> node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"

events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: spawn /usr/local/bin/node ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:229:19)
    at onErrorNT (internal/child_process.js:406:16)
    at process._tickCallback (internal/process/next_tick.js:63:19)
Emitted 'error' event at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:235:12)
    at onErrorNT (internal/child_process.js:406:16)
    at process._tickCallback (internal/process/next_tick.js:63:19)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! dotnet-2.0.0@1.4.4 postinstall: `node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the dotnet-2.0.0@1.4.4 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/vagrant/.npm/_logs/2018-08-01T15_02_14_622Z-debug.log`

**Comment by Zerryth** [comment]
@ponarunkumar 
@jwiley84 and I have used the command "`sudo npm install luisgen -g`" without being able to repro the issue

Going to investigate how else there may be discrepancies between the results you and @disophisis are having versus the ones jwiley & I are receiving

**Comment by jwiley84** [comment]
I have another idea. Investigating.

**Comment by jwiley84** [comment]
I am still investigating the matter, however, I believe the reason I ran into no issues using 'sudo npm install luisgen -g' on either my xenial or my bionic is because I was using dotnet 2.1. Further attempts to downgrade my dotnet to dotnet-2.0 and repro this have me running into a separate issue that will take a bit longer to correct ("no package dotnet-2.0.0 found" during installation). As soon as I am able to successfully install dotnet-2.0.0 and attempt another repro, I will update again.

**Comment by tsuwandy** [comment]
Sorry, just noticed this thread.  @ponarunkumar, did you follow instructions on this page to pre-install .NET Core on Ubuntu?
https://docs.microsoft.com/en-us/dotnet/core/linux-prerequisites?tabs=netcore2x

**Comment by Zerryth** [comment]
@ponarunkumar 

@jwiley84 and I have finally successfully repro'd the error you have gotten on 2 different Ubuntu servers:

- one with updated .NET Core
- one without 

and have both received the same error.

Continuing to investigate further

**Comment by Zerryth** [comment]
@vishwacsena added

Discussion

-  notes: unable to install .NET Core SDK 2.0.0 on Ubuntu 18.04 because 2.0.0 is unavailable to 18.04
- the error reads something along the lines of needing dotnet  core 2.0.0
- however it does look like in [install links](https://docs.microsoft.com/en-us/dotnet/core/linux-prerequisites?tabs=netcore2x#install-net-core-for-supported-red-hat-enterprise-linux-rhel-versions) that Tien provided, .NET Core 2.1.105+ SDK is available to Ubuntu 18.04

**Comment by vishwacsena** [comment]
Thanks @Zerryth. 

@chrimc62; @tsuwandy - the issue seems to be that we require 2.0.0 which does not support several linux distributions. Unless there is a hard dependency to 2.0.0, can we update both dispatch and luisgen to support the latest dotnet core v2.1.200? This would enable more customers to use the tools on a wider array of supported distributions.

**Comment by tsuwandy** [comment]
@Zerryth, I just checked this page https://docs.microsoft.com/en-us/dotnet/core/linux-prerequisites?tabs=netcore2x#install-net-core-for-supported-red-hat-enterprise-linux-rhel-versions and it has links to install .NET Core 2.0 runtime/SDK.   Did you try using one of those .NET Core 2.0 links for Ubuntu 18.04? SDK is actually not needed, only the runtime.   I'll try updating to .NET Core 2.1 regardless.

**Comment by Zerryth** [comment]
@tsuwandy regarding runtime - ah, I see.

Looking into it, it looks like Ubuntu 18.04 is avaiable on .NET Core Runtimes 2.0.7+ (i.e. not 2.0.5, 2.0.6)--if that matters at all.

In any case, just now I went ahead and installed Runtime 2.0.7 and then afterwards tried installing .NET SDK 2.1.105 (already had it) and got "Failed at the dotnet-2.0.0@1.4.4 postinstall script 'node -e ...." error

installs/upgrades + error below

```
ashleyho@AshleyHoUbuntu2:~$ wget -q https://packages.microsoft.com/config/ubuntu/18.04/packages-microsoft-prod.deb
ashleyho@AshleyHoUbuntu2:~$ sudo dpkg -i packages-microsoft-prod.deb
(Reading database ... 98515 files and directories currently installed.)
Preparing to unpack packages-microsoft-prod.deb ...
Unpacking packages-microsoft-prod (1.0-ubuntu18.04.1) over (1.0-ubuntu18.04.1) ...
Setting up packages-microsoft-prod (1.0-ubuntu18.04.1) ...
ashleyho@AshleyHoUbuntu2:~$ sudo apt-get install apt-transport-https
Reading package lists... Done
Building dependency tree
Reading state information... Done
apt-transport-https is already the newest version (1.6.3).
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
ashleyho@AshleyHoUbuntu2:~$ sudo apt-get update
Hit:1 https://packages.microsoft.com/repos/microsoft-ubuntu-bionic-prod bionic InRelease
Hit:2 http://azure.archive.ubuntu.com/ubuntu bionic InRelease
Get:3 http://azure.archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:4 http://azure.archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Get:5 http://azure.archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [277 kB]
Get:6 http://azure.archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [157 kB]
Get:7 http://security.ubuntu.com/ubuntu bionic-security InRelease [83.2 kB]
Fetched 680 kB in 1s (892 kB/s)
Reading package lists... Done
ashleyho@AshleyHoUbuntu2:~$ sudo apt-get install dotnet-hosting-2.0.7
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following NEW packages will be installed:
  dotnet-hosting-2.0.7
0 upgraded, 1 newly installed, 0 to remove and 1 not upgraded.
Need to get 2336 B of archives.
After this operation, 8192 B of additional disk space will be used.
Get:1 https://packages.microsoft.com/repos/microsoft-ubuntu-bionic-prod bionic/main amd64 dotnet-hosting-2.0.7 amd64 2.0.7-1 [2336 B]
Fetched 2336 B in 0s (10.4 kB/s)
Selecting previously unselected package dotnet-hosting-2.0.7.
(Reading database ... 98515 files and directories currently installed.)
Preparing to unpack .../dotnet-hosting-2.0.7_2.0.7-1_amd64.deb ...
Unpacking dotnet-hosting-2.0.7 (2.0.7-1) ...
Setting up dotnet-hosting-2.0.7 (2.0.7-1) ...
ashleyho@AshleyHoUbuntu2:~$ sudo apt-get install apt-transport-https
Reading package lists... Done
Building dependency tree
Reading state information... Done
apt-transport-https is already the newest version (1.6.3).
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
ashleyho@AshleyHoUbuntu2:~$ sudo apt-get  update
Hit:1 https://packages.microsoft.com/repos/microsoft-ubuntu-bionic-prod bionic InRelease
Hit:2 http://azure.archive.ubuntu.com/ubuntu bionic InRelease
Hit:3 http://azure.archive.ubuntu.com/ubuntu bionic-updates InRelease
Hit:4 http://azure.archive.ubuntu.com/ubuntu bionic-backports InRelease
Hit:5 http://security.ubuntu.com/ubuntu bionic-security InRelease
Reading package lists... Done
ashleyho@AshleyHoUbuntu2:~$ sudo apt-get install dotnet-sdk-2.1.105
Reading package lists... Done
Building dependency tree
Reading state information... Done
dotnet-sdk-2.1.105 is already the newest version (2.1.105-1).
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
ashleyho@AshleyHoUbuntu2:~$ sudo npm install luisgen -g
/usr/local/bin/luisgen -> /usr/local/lib/node_modules/luisgen/bin/luisgen.js                                                                                                                 ░░░╟
                                                                                                                                                                                             ░░░╟
> dotnet-2.0.0@1.4.4 postinstall /usr/local/lib/node_modules/luisgen/node_modules/dotnet-2.0.0
> node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"

events.js:183                                                                                                                                                                                ░░░╟
      throw er; // Unhandled 'error' event
      ^

Error: spawn /usr/bin/node ENOENT
    at _errnoException (util.js:1022:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:190:19)
    at onErrorNT (internal/child_process.js:372:16)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! Linux 4.15.0-1018-azure                                                                                                                                                             ░░░╟
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "luisgen" "-g"
npm ERR! node v8.10.0
npm ERR! npm  v3.5.2
npm ERR! code ELIFECYCLE

npm ERR! dotnet-2.0.0@1.4.4 postinstall: `node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the dotnet-2.0.0@1.4.4 postinstall script 'node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the dotnet-2.0.0 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node -e "/*PostInstall: Installs platform-specific .NET framework */try{require('./dist/app.js')}catch(e){}"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs dotnet-2.0.0
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls dotnet-2.0.0
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ashleyho/npm-debug.log
```

**Comment by sgellock** [comment]
Blocked on .net core supporting this specific distro

**Comment by vishwacsena** [comment]
@tsuwandy; @chrimc62 ; @Zerryth  - can this be resolved? AFAIK, we just need to update readme for 2.1 install instructions - right?

**Comment by tsuwandy** [comment]
Yes, we have now switched to .NET Core 2.1 for both LuisGen and Dispatch CLI.   .NET Core 2.1 is supported in Ubuntu 18.04 distro.  @ponarunkumar, could you try again and reactivate otherwise?  I have verified that latest Dispatch works in Ubuntu 18.04 myself.
